src/agent/coverage/src/block/windows.rs

```diff
@@ -7,162 +7,257 @@ use std::time::{Duration, Instant};
-            bp_to_block,
-            coverage,
-            max_duration,
+            recorder,
             started,
             timed_out,
+            timeout,
         }
     }
 
-    pub fn pc(&self, dbg: &mut Debugger) -> Result<(u64, Option<SymInfo>)> {
-        use iced_x86::Register::RIP;
-
-        let pc = dbg.read_register_u64(RIP)?;
-        let sym = dbg.get_symbol(pc).ok();
-
-        Ok((pc, sym))
+    pub fn time_out(&self) -> bool {
+        self.timed_out
     }
 
-    fn add_module(&mut self, dbg: &mut Debugger, module: &Module) {
-        let bitset = crate::pe::process_image(module.path(), false);
-        let bitset = match bitset {
-            Ok(bitset) => bitset,
-            Err(err) => {
-                // If we can't add the module, continue debugging.
-                // We don't expect to have symbols for every module.
-                log::warn!(
-                    "cannot record coverage for module = {}, err = {}",
-                    module.path().display(),
-                    err,
-                );
-                return;
-            }
-        };
+    pub fn timeout(&self) -> Duration {
+        self.timeout
+    }
 
-        let module_coverage = ModuleCoverageBlocks::new(module.path(), module.name(), bitset);
+    pub fn run(&mut self, cmd: Command) -> Result<()> {
+        let (mut dbg, _child) = Debugger::init(cmd, self)?;
+        dbg.run(self)?;
+        Ok(())
+    }
 
-        let m = self.coverage.add_module(module_coverage);
-        let module_coverage = &self.coverage.modules()[m];
-        for (b, block) in module_coverage.blocks().iter().enumerate() {
-            let bp =
-                dbg.register_breakpoint(module.name(), block.rva() as u64, BreakpointType::OneTime);
-            self.bp_to_block.insert(bp, (m, b));
+    fn on_poll(&mut self, dbg: &mut Debugger) {
+        if !self.timed_out && self.started.elapsed() > self.timeout {
+            self.timed_out = true;
+            dbg.quit_debugging();
         }
-
-        log::debug!(
-            "inserted {} breakpoints for module {}",
-            module_coverage.blocks().len(),
-            module.path().display(),
-        );
     }
 
     fn stop(&self, dbg: &mut Debugger) {
         dbg.quit_debugging();
     }
+}
+
+#[derive(Debug)]
+pub struct Recorder<'a> {
+    breakpoints: Breakpoints,
 
-    fn handle_on_create_process(&mut self, dbg: &mut Debugger, module: &Module) -> Result<()> {
-        dbg.target().sym_initialize()?;
+    // Reference to allow in-memory reuse across runs.
+    cache: &'a mut ModuleCache,
 
-        log::info!(
-            "exe loaded: {}, {} bytes",
-            module.path().display(),
-            module.image_size(),
-        );
+    // Note: this could also be a reference to enable reuse across runs, to
+    // support implicit calculation of total coverage for a corpus. For now,
+    // assume callers will merge this into a separate struct when needed.
+    coverage: CommandBlockCov,
```

Comment:
It seems reasonable for `Recorder` to **not** do any merging.

I do wonder though what mechanism we would use to skip collecting coverage on already hit blocks. This would speed up collecting total coverage for a corpus and would also speed up fuzzing iterations when using radamsa.

Reply:
Good point. I was thinking we'll want the `ModuleCache` itself to be append-only. However, we could still add a second optional reference to an existing (total) `CommandBlockCov`, and when setting breakpoints, we could skip registration if a hit already occurs in the existing cov. I think this may warrant separate profiling, however.